### PR TITLE
Core/ICC: Lady Deathwhisper

### LINKED ISSUE #761: Lady Deathwhisper Major Bugs
**Description**: Lady Deathwhisper currently has many Bugs during both Phases.

**Current behaviour**: 
Phase 1
- Empowered Cult Adherents are NOT immune to interrupt effects.
- You can "interrupt" Dark Martyrdom casts with Stuns and with knockback effects such as Thunderstorm / Typhoon.
- If Dark Martyrdom cast is successful, the Reanimated add spawns with a many seconds delay.

Phase 2
- Adds dying from Dark Martyrdom in Phase 2 are NOT reviving as Reanimated.

**Expected behaviour**: 
- Empowered Cult Adherents definitely should be immune to interrupts.
- The only way to "interrupt" Dark Martyrdom is killing the mob before the cast is over. That means they should be immune to Stuns, and not suffering knockbacks.
- Adds dying from Dark Martyrdom should almost istantly spawn as Reanimated, even during Phase 2.

**Steps to reproduce the problem**:

1. Try to interrupt Empowered Cult Adherents and notice it actually works.
2. Try to stun adds during Dark Martyrdom cast and see it actually stuns them.
3. Notice how adds who could successfully cast Dark Martyrdom may take many seconds to spawn as Reanimated.
4. During Phase 2 adds are not spawning as Reanimated AT ALL.

**Branch(es)**: Master

**AC hash/commit**:  29f4a1e090338bc033ebdecb0d71ee45cde6f66f

**Operating system**:  Linux

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/54023773-lady-deathwhisper-major-bugs?utm_campaign=plugin&utm_content=tracker%2F40032087&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F40032087&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### LINKED ISSUE #1538: [ICCMOB][PVE][SPELL] Lady Deathwhisper- Second Phase "Dark Martyrdom"
##### SMALL DESCRIPTION:
During the second phase of the Boss of Icc:
The Lady Deathwisper ( https://www.wowhead.com/npc=36855/lady-deathwhisper ) adds that spawn
( https://www.wowhead.com/npc=37949/cult-adherent -https://www.wowhead.com/npc=37890/cult-fanatic ) during the second phase should not use the spell Dark Martyrdom
( https://www.wowhead.com/spell=70903/dark-martyrdom#used-by-npc ADHERENT )
( https://www.wowhead.com/spell=71236/dark-martyrdom FANATIC )

##### EXPECTED BLIZZLIKE BEHAVIOUR:
The Lady Deathwisper ( https://www.wowhead.com/npc=36855/lady-deathwhisper ) adds that spawn
( https://www.wowhead.com/npc=37949/cult-adherent -https://www.wowhead.com/npc=37890/cult-fanatic ) during the second phase should not use the spell Dark Martyrdom
( https://www.wowhead.com/spell=70903/dark-martyrdom#used-by-npc ADHERENT )
( https://www.wowhead.com/spell=71236/dark-martyrdom FANATIC )

Tankspot Offy raid - same behaviour 10h and 25 h  https://youtu.be/W64bCmigL1Q 

##### CURRENT BEHAVIOUR:
During the First Phase this adds ( THE SAME ADDS) works but during the second phase when the Lady Deathwhisper have no more the mana shield this adds don't works as intended

##### STEPS TO REPRODUCE THE PROBLEM:
1. Go Inside Icc  ( 10h-25h) and after have cleaned all the room of the Lady Deathwhisper start the fight
2. see that in the first phase this adds works correctly and use the Dark Martyrdom how should be 
3. see that in the second phase this adds use that spell the Dark Martyrdom that they should not use 

##### EXTRA NOTES:
HIGH PRIO !

##### AC HASH/COMMIT:

https://github.com/azerothcore/azerothcore-wotlk/commit/44272f169c49f20c79b78aeea634f3bc7c2dafd1

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/70722310-iccmob-pve-spell-lady-deathwhisper-second-phase-dark-martyrdom?utm_campaign=plugin&utm_content=tracker%2F40032087&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F40032087&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by Nefertumm**
##### CHANGES PROPOSED:

-  Added missing overrides and changed NULL for nullptr.
-  Fix timers (this needs atention, it might be wrong)
-  No more Dark Martyrdom being casted in phase 2.
-  Added temporal immunities to certain spell mechanics to avoid interrupt Dark Martyrdom cast.
-  Added the correct auras to reanimated cultists (fanatic's/adherent's determination).
-  Avoid spam of Dark Martyrdom.
-  Lady Deathwhisper should be immune to taunt on heroic.

###### ISSUES ADDRESSED:

Closes #1538 
Closes #761 
Maybe updates #1532 

##### TESTS PERFORMED:

Compiled and tested. Need more tests though.

##### HOW TO TEST THE CHANGES:

Go to icc, kill Lord Marrowgar and go to Lady Deathwhisper. Play with mobs during first phase, make sure you can't interrupt Dark Martyrdom's cast, etc... Detailed above.

##### Target branch(es):

Master

### CODE REVIEW

**Comment by Nefertumm** [comment]
I miss the sql file, I'm gonna upload it when i get up.

File: src/server/scripts/Northrend/IcecrownCitadel/boss_lady_deathwhisper.cpp [hunk 1]
```diff
@@ -375,31 +403,27 @@ class boss_lady_deathwhisper : public CreatureScript
                     case EVENT_SPELL_SHADOW_BOLT:
                         if (Unit* target = SelectTarget(SELECT_TARGET_RANDOM))
                             me->CastSpell(target, SPELL_SHADOW_BOLT, false);
-                        events.RepeatEvent(urand(5000, 8000));
+                        events.RepeatEvent(urand(2450, 3600));
```
Comment (GitVerge) [inline src/server/scripts/Northrend/IcecrownCitadel/boss_lady_deathwhisper.cpp]: I am not sure about this timer, it feels like 2s 2s 2s 2s, stop, 2s, 2s... to me

File: src/server/scripts/Northrend/IcecrownCitadel/boss_lady_deathwhisper.cpp [hunk 2]
```diff
@@ -671,48 +666,61 @@ class npc_cult_fanatic : public CreatureScript
...
+                case SPELL_DARK_MARTYRDOM_FANATIC:
+                case SPELL_DARK_MARTYRDOM_FANATIC_10H:
+                case SPELL_DARK_MARTYRDOM_FANATIC_25N:
+                case SPELL_DARK_MARTYRDOM_FANATIC_25H:
+                    ApplyMechanicImmune(me, false);
+                    me->GetMotionMaster()->MovementExpired();
+                    me->GetMotionMaster()->MoveIdle();
+                    me->StopMoving();
+                    Reset();
+                    me->UpdateEntry(NPC_REANIMATED_FANATIC);
```
Comment (pklloveyou) [inline src/server/scripts/Northrend/IcecrownCitadel/boss_lady_deathwhisper.cpp]: Explosion effects are not shown here

File: src/server/scripts/Northrend/IcecrownCitadel/boss_lady_deathwhisper.cpp [hunk 3]
```diff
@@ -671,48 +666,61 @@ class npc_cult_fanatic : public CreatureScript
...
+                    me->CastSpell(me, SPELL_PERMANENT_FEIGN_DEATH);
```
Comment (pklloveyou) [inline src/server/scripts/Northrend/IcecrownCitadel/boss_lady_deathwhisper.cpp]: Explosion effects are not shown here

File: src/server/scripts/Northrend/IcecrownCitadel/boss_lady_deathwhisper.cpp [hunk 4]
```diff
@@ -671,48 +666,61 @@ class npc_cult_fanatic : public CreatureScript
...
+                    me->SetFlag(UNIT_DYNAMIC_FLAGS, UNIT_DYNFLAG_DEAD);
```
Comment (pklloveyou) [inline src/server/scripts/Northrend/IcecrownCitadel/boss_lady_deathwhisper.cpp]: Explosion effects are not shown here

**Comment by Nefertumm** [comment]
What do you mean with you cant use this feature?

**Comment by pklloveyou** [comment]
The online translation is incorrect
These few lines of code will cause NPCS to not display explosions

**Comment by comix1988** [comment]
today I tested the lady at 10 normal icc:

https://youtu.be/lUeNRXXAKNQ

- sometimes when the npc cult adherent gets stunned it freezes (0:45 min) ----> to fix
- no NPC spawned in phase 2; ----> GOOD
- reanimated cultists have a correct aura; ----> GOOD
- interrupt Dark Martyrdom cast is not possible; ----> GOOD
-timer of adds weaves work perfectly until the third,then delay a few seconds(3:06 and 4:08 mins)-> to fix
- when the cast of Dark Martyrdom is completed the npc explosion isn't seen --> fix visual

tomorrow i test other difficulties
@heisenberg520 see other bugs?

tested on Ubuntu 18.04.2 LTS

**on behalf of @maelstromserver**

**Comment by comix1988** [comment]
Lady 10H:

https://youtu.be/sGOuKUBK54Q

- the spawn time of the add should be fixed (the second wave anticipates 10 sec);
- in the second phase it doesn't spawn Vengeful Spirits;
- when the cast of Dark Martyrdom is completed the npc explosion isn't seen --> fix visual
- everything else is OK

tested on Ubuntu 18.04.2 LTS

**on behalf of @MaelstromServer**

**Comment by comix1988** [comment]
lady 25n:

https://youtu.be/UOo3RNzp79U

- timer of adds weaves work perfectly until the third,then delay a few seconds(3:06 and 4:08 mins)-> to fix
- when the cast of Dark Martyrdom is completed the npc explosion isn't seen --> fix visual

**Comment by comix1988** [comment]
lady 25h:

https://youtu.be/_E2oIv13GrQ

- the spawn time of the add should be fixed (the second wave anticipates 10 sec);
- when the cast of Dark Martyrdom is completed the npc explosion isn't seen --> fix visual

**Comment by comix1988** [comment]
definitely: @Nefertumm 

- in every the difficulties fix the visual bug of the explosion of the mob that has cast "dark martyrdom";
- check why sometimes the npc "cult adherent" freezes;
- fix the add waves: in normal mode from the third wave they delay by 2 sec;
                                in hero mode anticipate by about 12 seconds from the second wave;

everything else is ok for me. I hope I was precise in the explanation.

P.S: in the videos of the hero mode the vangeful spirits are not seen but they work, I tested them with 5 pg instead of 2 pg.

tested on Ubuntu 18.04.2 LTS

**on behalf of @MaelstromServer**

**Comment by Nefertumm** [comment]
WELL, I FUCKED UP BOYS

**Pull request closed without merging**